ocean_lib/models/datatoken_enterprise.py

```diff
@@ -18,37 +21,37 @@ def buy_from_fre_and_order(
         self,
         consumer: str,
         service_index: int,
-        provider_fee_address: str,
-        provider_fee_token: str,
-        provider_fee_amount: Union[int, str],
-        v: int,
-        r: Union[str, bytes],
-        s: Union[str, bytes],
-        valid_until: int,
-        provider_data: Union[str, bytes],
+        provider_fees: dict,
         consume_market_order_fee_address: str,
         consume_market_order_fee_token: str,
         consume_market_order_fee_amount: int,
```

Comment:
This is best as some sort of "fees" class. It's a recurring pattern that holds a 3-tuple of fees info: address, token, amount. Not just for consume_market_order_fee, but other things. Call it e.g. "FeeGroup".

I recommend: create such a fees class, and use it liberally to replace input arguments (like here), and to replace raw "getter" calls to smart contracts such as DT.getPublishingMarketFee().

You can look to fixed_rate_exchange.py::ExchangeDetails class, and fixed_rate_exchange.py::FeesInfo class for inspiration. E.g. copy & paste one of those, then chop it down into holding 3 attributes ( address, token, amount).

Reply:
I would do that in a next iteration. I can mark this PR as "work on ..." instead of "closes ..." and I propose we merge it in (of course after fixing the other comments), then I'll do the fee grouping in a different PR to help with conflict management. Would that work?

Reply:
Yes! 

I'd actually suggested the exact same thing in slack :)

Reply:
I see you've got partway here (for provider fees), by having a "provider_fees" dict. It already greatly simplified things.

But you'll notice that by having the dict
- it doesn't have an easy way to create defaults
- when grabbing parameter values it's slightly sloppy. E.g. `provider_fees["providerFeeAddress"]`

If you create a class from it, ie have a `ProviderFees` class, then you will be able to resolve both
- defaults are easy
- and cleaner to grab parameter values, e.g. `provider_fees.address` (`provider_fees` is an instance of `ProviderFee`)

Also, you'll want a class for `ConsumeMarketOrderFees` too. Whereas `ProviderFees` will have 6+ attributes, `ConsumeMarketOrderFees` will just have 3 or so.

Like mentioned above, you can look to fixed_rate_exchange.py::ExchangeDetails class, and fixed_rate_exchange.py::FeesInfo class for inspiration. 

(This is probably the largest single remaining thing to do. And it will help a lot.)

Reply:
I don't agree with changing provider fees to a class. We get them from provider as a dict, through the response itself, so we can use them straight through. There is no case where we "build" the provider fees except for tests where we emulate the provider. I will do it if needed but I think it is overengineering in this case

Reply:
OK. Those are good arguments to keep provider fees as a dict. Fine, then:)

What about ConsumeMarketOrderFees?

Reply:
UPDATE: we'll put ConsumeMarketOrderFees etc into another github issue

Reply:
Github issue: https://github.com/oceanprotocol/ocean.py/issues/1222